compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -5636,6 +5662,97 @@ private STNode parseObjectMethodOrField(SyntaxKind nextTokenKind, SyntaxKind nex
         return parseObjectMethodOrField(solution.tokenKind, nextTokenKind, metadata, visibilityQualifier);
     }
 
+    /**
+     * Parse function qualifiers.
+     *
+     * @return Parsed node
+     */
+    private STNode parseFunctionQualifiers(ParserRuleContext context, List<STNode> qualifierList) {
+        STToken nextToken = peek();
+        return parseFunctionQualifiers(nextToken.kind, context, qualifierList);
+    }
+
+    private STNode parseFunctionQualifiers(SyntaxKind tokenKind, ParserRuleContext context,
+                                               List<STNode> qualifierList) {
+        while (!isEndOfFunctionQualifiers(tokenKind)) {
+            STNode qualifier;
+            switch (tokenKind) {
+                case REMOTE_KEYWORD:
+                    qualifier = parseRemoteKeyword();
+                    break;
+                case TRANSACTIONAL_KEYWORD:
+                    qualifier = parseTransactionalKeyword();
+                    break;
+                case RESOURCE_KEYWORD:
+                    qualifier = parseResourceKeyword();
+                    break;
+                default:
+                    Solution solution = recover(peek(), context, context, qualifierList);
+
+                    // If the parser recovered by inserting a token, then try to re-parse the same
+                    // rule with the inserted token. This is done to pick the correct branch
+                    // to continue the parsing.
+                    if (solution.action == Action.REMOVE) {
+                        return solution.recoveredNode;
```

Comment:
Can returning immediately be a problem?
eg:
```ballerina
transactional xyz resource function foo() {
}
```
Here `xyz` is an invalid token. After recovering from it, shouldn't the loop continue?

Reply:
No, In the case of invalid token after recovery it goes to resume parsing. Then it parses the rest of the qualifiers and return here so we return immediately right?

Reply:
ah ok. Have any tests covering that?

Reply:
Yes already added recovery test cases for fuction-def, object-method and service resource method. Those tests contains your example as well.